pkg/volume/persistent_claim/persistent_claim.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/gce_pd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/glusterfs"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/host_path"
+	"github.com/golang/glog"
+)
+
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&persistentClaimPlugin{nil}}
+}
+
+type persistentClaimPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &persistentClaimPlugin{}
+
+const (
+	persistentClaimPluginName = "kubernetes.io/persistent-claim"
+)
+
+func (plugin *persistentClaimPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *persistentClaimPlugin) Name() string {
+	return persistentClaimPluginName
+}
+
+func (plugin *persistentClaimPlugin) CanSupport(spec *api.Volume) bool {
+	return spec.PersistentVolumeClaimVolumeSource != nil
+}
+
+func (plugin *persistentClaimPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+	glog.V(5).Infof("Creating builder for %+v\n", spec)
+	claimName := spec.PersistentVolumeClaimVolumeSource.ClaimName
+	podNamespace := podRef.Namespace
+	claim, err := plugin.host.GetKubeClient().PersistentVolumeClaims(podNamespace).Get(claimName)
+
+	glog.V(5).Infof("Using claim: %+v\n", claim)
+
+	if err != nil {
+		glog.Errorf("Error finding claim in namespace  PersistentVolume by ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+		return nil, err
+	}
+
+	if claim != nil {
+		pv, err := plugin.host.GetKubeClient().PersistentVolumes().Get(claim.Status.VolumeRef.Name)
+		if err != nil {
+			glog.Errorf("Error finding bound PersistentVolume by ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+			return nil, err
+		}
+
+		glog.V(5).Infof("Using persistent volume: %+v\n", pv)
+		volPlugin := getVolumePlugin(pv.Spec)
+
+		if volPlugin == nil {
+			glog.Errorf("No plugin found for PersistentVolume: %+v\n", pv)
+			return nil, err
+		}
+
+		glog.V(5).Infof("Using plugin: %s\n", volPlugin.Name())
+		volume, err := toVolume(pv)
+		if err != nil {
+			glog.Errorf("Error getting VolumeSource for %+v: \n", pv.Spec.PersistentVolumeSource)
+			return nil, err
+		}
+
+		volPlugin.Init(plugin.host)
+		builder, err := volPlugin.NewBuilder(volume, podRef)
+
+		glog.V(5).Infof("Returning builder: %s\n", builder.GetPath())
+
+		if err != nil {
+			glog.Errorf("Error creating builder for ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+			return nil, err
+		}
+
+		if builder != nil {
+			return builder, nil
+		}
+	}
+
+	return nil, fmt.Errorf("No builder found for volume %+v\n", spec)
+}
+
+func (plugin *persistentClaimPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return nil, fmt.Errorf("This will never be called directly.  Volumes backing a PV have their own cleaners and Kubelet reconciles volumes in its syncLoop.")
+}
+
+func toVolume(pv *api.PersistentVolume) (*api.Volume, error) {
```

Comment:
Then you have to actively keep that structure in sync.  By just embedding the two options you don't ever have to update that struct ever again.  Only kubelet and plugins ever see that struct. And who knows, once we have API plugins, we'll have to come back to this anyway.